### PR TITLE
detect: absent keyword to test absence of sticky buffer

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2224

Describe changes:
- detect: adds `absent` keyword to match on absent buffer

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1957

https://github.com/OISF/suricata/pull/11459 with code review taken into account

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 21549

**Review by inashivb — APPROVED** [review]
LGTM 🚀

**Comment by jufajardini** [comment]
Docs look good to me :)

**Comment by catenacyber** [comment]
@victorjulien what do you think ?

File: src/detect-file-data.c [hunk 1]
```diff
@@ -400,7 +400,15 @@ uint8_t DetectEngineInspectFiledata(DetectEngineCtx *de_ctx, DetectEngineThreadC
 
     AppLayerGetFileState files = AppLayerParserGetTxFiles(f, txv, flags);
     FileContainer *ffc = files.fc;
-    if (ffc == NULL) {
+    if (ffc == NULL || ffc->head == NULL) {
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-file-data.c:403]: probably should just return `DETECT_ENGINE_INSPECT_SIG_CANT_MATCH_FILES` for `ffc==NULL` and only check `ffc->head` and such after that check

File: src/detect-http-header.c [hunk 2]
```diff
@@ -209,14 +214,8 @@ static uint8_t DetectEngineInspectBufferHttpHeader(DetectEngineCtx *de_ctx,
         return DETECT_ENGINE_INSPECT_SIG_MATCH;
     }
 end:
-    if (flags & STREAM_TOSERVER) {
```
Comment (victorjulien) [inline src/detect-http-header.c:212]: this change looks like it possibly changes behavior? No explicit HTTP1 anymore... probably deserves its own commit and explanation

File: src/detect-isdataat.c [hunk 3]
```diff
@@ -56,11 +57,97 @@ static DetectParseRegex parse_regex;
...
+        SCLogError("unspected buffer for absent keyword");
+        return -1;
+    }
+    const DetectBufferType *b = DetectEngineBufferTypeGetById(de_ctx, s->init_data->list);
+    if (!b || b->frame) {
+        SCLogError("absent does not work with frames");
+        return -1;
+    }
+    if (s->init_data->curbuf->tail != NULL) {
+        SCLogError("absent must come first right after buffer");
```
Comment (victorjulien) [inline src/detect-isdataat.c:100]: what is the relationship with buffers, transforms and absent? Do we have tests?

File: src/detect-isdataat.c [hunk 4]
```diff
@@ -56,11 +57,97 @@ static DetectParseRegex parse_regex;
...
+                if (has_absent && (cd->flags & DETECT_CONTENT_FAST_PATTERN)) {
+                    SCLogError("signature can't have absent and fast_pattern on the same buffer");
+                    return false;
+                }
+            }
+        }
+    }
+
+    if (only_absent && has_other) {
+        SCLogError("signature can't have a buffer only absent and tested otherwise");
```
Comment (victorjulien) [inline src/detect-isdataat.c:142]: sentence is a bit unclear

File: src/detect-isdataat.c [hunk 5]
```diff
@@ -56,11 +57,97 @@ static DetectParseRegex parse_regex;
...
+        return false;
+    } else if (has_absent && !only_absent && !has_other) {
+        SCLogError("signature with absent: or_else expects something else to test on");
```
Comment (victorjulien) [inline src/detect-isdataat.c:145]: something else should probably be something like "other match conditions"

File: src/detect-isdataat.h [hunk 6]
```diff
@@ -34,7 +34,14 @@ typedef struct DetectIsdataatData_ {
     uint8_t flags; /* isdataat options*/
 } DetectIsdataatData;
 
+typedef struct DetectAbsentData_ {
+    // absent or something else (or only absent)
```
Comment (victorjulien) [inline src/detect-isdataat.h:38]: please use doxygen format

File: src/detect.h [hunk 7]
```diff
@@ -432,6 +432,8 @@ typedef struct DetectEngineAppInspectionEngine_ {
     uint8_t id;     /**< per sig id used in state keeping */
     bool mpm;
     bool stream;
+    /** will match on a NULL buffer, cf absent keyword */
```
Comment (victorjulien) [inline src/detect.h:435]: `NULL buffer (so an absent buffer)`

**Review by victorjulien — CHANGES_REQUESTED** [review]
Meant to use request-changes

File: src/detect-engine-content-inspection.c [hunk 8]
```diff
@@ -382,6 +382,14 @@ static int DetectEngineContentInspectionInternal(DetectEngineThreadCtx *det_ctx,
                     prev_offset);
         } while(1);
 
+    } else if (smd->type == DETECT_ABSENT) {
+        const DetectAbsentData *id = (DetectAbsentData *)smd->ctx;
+        if (!id->or_else) {
+            // we match only on absent buffer
+            goto no_match;
+        } else {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-engine-content-inspection.c:390]: ok

File: src/detect-isdataat.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-isdataat.c:100]: Do you mean that we should explicitly forbid transforms on buffers where we have `absent` ?

File: src/detect-isdataat.c (hunk 4 above)
Comment (catenacyber) [reply src/detect-isdataat.c:142]: rephrasing

File: src/detect.h (hunk 7 above)
Comment (catenacyber) [reply src/detect.h:435]: ok

File: src/detect-isdataat.h (hunk 6 above)
Comment (catenacyber) [reply src/detect-isdataat.h:38]: ok

File: src/detect-http-header.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-http-header.c:212]: > this change looks like it possibly changes behavior?

It should not :
- DetectEngineInspectBufferHttpHeader is only used with ALPROTO_HTTP1
- engine->progress should be HTP_REQUEST_HEADERS or HTP_RESPONSE_HEADERS based on the direction 

> probably deserves its own commit and explanation

ok

File: src/detect-file-data.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-file-data.c:403]: If a protocol does not support files in that direction (like SMTP), `DetectEngineInspectFiledata` is only called with the right direction (tested with debug print on smtp-attachment-md5 with added rule `alert smtp any any -> any any (file.data; absent; sid:5; rev:1;)`)

[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-file-data.c:403]: > probably should just return DETECT_ENGINE_INSPECT_SIG_CANT_MATCH_FILES for ffc==NULL and only check ffc->head and such after that check

ok

[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-file-data.c:403]: Tested with 
```
alert http2 any any -> any any (flow: to_server; file.data; absent; http2.frametype:HEADERS; sid:5; rev:1;)
alert http2 any any -> any any (flow: to_client; file.data; absent; http2.frametype:HEADERS; sid:6; rev:1;)
```
in http2-basic SV test, to see that we get alerts for GET requests which do not send file to server

**Comment by catenacyber** [comment]
Next in https://github.com/OISF/suricata/pull/11906

File: src/detect-file-data.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-file-data.c:403]: There is also already detect-absent-file-multi in SV PR
